blsftcosi/README.md

```diff
@@ -0,0 +1,73 @@
+[Cothority](../README.md) ::
+[Building Blocks](../doc/BuildingBlocks.md) ::
+Fault Tolerant Collective Signing
+
+# Fault Tolerant Collective Signing
+
+This package provides functionality to request and verify collective signatures
+as well as run a standalone server for handling collective signing requests.
+It is a fault tolerant version of CoSi, implemented in
+[cosi](../cosi/README.md) package.
+
+## Research Paper
+
+It is the basis for the ByzCoinX protocol. For further background and technical
+details, please refer to the
+[research paper](https://eprint.iacr.org/2017/406.pdf).
+
+You can read more about BLS signature by refering to
+[the paper](https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html).
+
+## Description
+
+The purpose of this work is to implement a robust and scalable consensus
+algorithm using ftCoSi protocol and handling some exceptions. The ftCoSi tree
+is a three level tree to make a compromise between the two-level tree, making
+the root-node vulnerable to DoS, and a more than three level tree, slowing the
+algorithm because of the RTT between the root node and the leaves.
+
+The tree is composed of a leader (root-node), and some groups of equal size,
+each having a sub-leader (second level nodes) and members (leaves). The group
+composition are defined by the leader.
+
+Ideally, we want to handle non-responding nodes, no matter where they are
+in the tree. If a leaf is failing, then it is ignored in the ftCoSi commitment.
+If a sub-leader is non-responding, then the leader (root node) recreates the
+group by selecting another sub-leader from the group members. And finally, if
+the leader is failing, the protocol restarts using another leader. At the
+moment, however, we only handle leaf and sub-leader failure.
+
+## Implementation
+The protocol has two messages: 
+- Announcement which is sent from the root down the tree and announce the
+proposal. 
+- Response which is sent back up to the root, containing the final aggregated
+signature, then used by the root to sign the proposal.
+
+The protocol uses four files: 
+- `struct.go` defines the messages sent around and the protocol constants.  
+- `protocol.go` defines the root node behavior.
+- `subprotocol.go` defines non-root nodes behavior.
+- `gen_tree.go` contains the function that generates trees.
+
+Under-the-hood, there are two protocols. A main protocol which only runs on
+the root node and a sub-protocol that runs on all nodes (including the
+root). The tree structure of the sub-protocol is illustrated below.
+
+```
+       root
+         |
+         |
+     sub-leader
+      /       \
+     /         \
+  leaf_1 ... leaf_n
+```
+
+Namely, if there are m sub-leaders, the root will run m sub-protocols. The
+sub- protocols do bulk of the work (collective signatures) and communicates
+the result to the main protocol via channels.
+
+- [BlsFtCosi CLI](blscosi/README.md) is a command line interface for interacting with ftCoSi
```

Comment:
why is the package/folder name called `blscosi` and not `blsftcosi`? maybe it's better to name it to `blsftcosicli` (it's a bit of a mouthful.. i'm open to suggestions). Also, "ftCoSi" need to be updated.

Reply:
but if we can't think of a better name then let's keep it as `blscosi` for the CLI

Reply:
Yeah I wasn't sure which name to use. I remove the _ft_  to make it shorter as it's not fundamental and wanted to avoid cli in the name as the package name is used as the executable name.

Reply:
Let's take feedback from @jeffallen and @ineiti for that

Reply:
Wondering whether we actually need a CLI for that. Probably for demo purposes it would be nice to have one. Some more suggestions:
- `blssign`
- `bfcsign` - BlsFtCosiSIGN
- `blscosi`

we could even remove the `ft` part entirely from the protocol, as this was mostly used to distinguish between the `cosi` and the `ftcosi`. But as there is no _non-fault-tolerant_ bls-cosi, we can remove it altogether and make the name a bit shorter.

My 500 Satoshis (2 cents)